internal/pprof/pprof.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pprof
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/pprof"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+// Server represents an HTTP server to serve Memory Profiling.
+// Note that this is only available in debug mode.
+type Server interface {
+	Start()
+	Stop()
+}
+
+type server struct {
+	log    logging.Logger
+	server *http.Server
+
+	port   int
+	router *mux.Router
+}
+
+// NewServer creates a new HTTP server for Memory Profiling.
+// It listens on a separate HTTP port and only available in
+// debug mode.
+func NewServer(log logging.Logger) Server {
+	return &server{
+		log:    log,
+		port:   9090,
+		router: mux.NewRouter(),
+	}
+}
+
+// Start memory profiling server
+func (s *server) Start() {
+	s.router.Handle("/debug/pprof/", http.HandlerFunc(pprof.Index))
+	s.router.Handle("/debug/pprof/cmdline", http.HandlerFunc(pprof.Cmdline))
+	s.router.Handle("/debug/pprof/profile", http.HandlerFunc(pprof.Profile))
+	s.router.Handle("/debug/pprof/symbol", http.HandlerFunc(pprof.Symbol))
+	s.router.Handle("/debug/pprof/trace", http.HandlerFunc(pprof.Trace))
+	s.router.Handle("/debug/pprof/goroutine", pprof.Handler("goroutine"))
+	s.router.Handle("/debug/pprof/heap", pprof.Handler("heap"))
+	s.router.Handle("/debug/pprof/threadcreate", pprof.Handler("threadcreate"))
+	s.router.Handle("/debug/pprof/block", pprof.Handler("block"))
```

Comment:
Any reason not to use the default mux? It will expose each of these for you if you do an anonymous pprof import (ref: https://golang.org/pkg/net/http/pprof/)

Reply:
The reason was essentially twofold 1) spin up a dedicated debug server on port `9090` and 2) pass additional profilers which don't exist in [`init()` of that package](https://golang.org/src/net/http/pprof/pprof.go#L79).

Reply:
And also what do you think of moving this to `crossplane-runtime` to be able to DRY it with providers as well (if they choose to utilize it)?

Reply:
@khos2ow adding this to crossplane-runtime feels pretty reasonable to me :)

Reply:
@hasheddan I've opened up https://github.com/crossplane/crossplane-runtime/pull/233 and after that gets merged we can proceed with this one with a mandatory change to `go.mod`.